File: scripts/levenshtein.py
def levenshtein_distance(word1, word2):
    len_word1 = len(word1) + 1
    len_word2 = len(word2) + 1

    # Initialize a matrix to store the distances
    matrix = [[0 for _ in range(len_word2)] for _ in range(len_word1)]

    # Initialize the first row and column of the matrix
    for i in range(len_word1):
        matrix[i][0] = i

    for j in range(len_word2):
        matrix[0][j] = j

    # Fill in the matrix
    for i in range(1, len_word1):
        for j in range(1, len_word2):
            cost = 0 if word1[i - 1] == word2[j - 1] else 1
            matrix[i][j] = min(
                matrix[i - 1][j] + 1,      # Deletion
                matrix[i][j - 1] + 1,      # Insertion
                matrix[i - 1][j - 1] + cost  # Substitution
            )

    # The bottom-right cell of the matrix contains the Levenshtein distance
    return matrix[-1][-1]

# Example usage
word1 = "em2in4ence"
word2 = "eminance"

distance = levenshtein_distance(word1, word2)
print(f"Levenshtein distance between {word1} and {word2}: {distance}")

File: scripts/levenshtein.js
function levenshteinDistance(word1, word2) {
    const lenWord1 = word1.length + 1;
    const lenWord2 = word2.length + 1;

    const matrix = Array.from({ length: lenWord1 }, () => Array(lenWord2).fill(0));

    for (let i = 0; i < lenWord1; i++) {
        matrix[i][0] = i;
    }

    for (let j = 0; j < lenWord2; j++) {
        matrix[0][j] = j;
    }

    for (let i = 1; i < lenWord1; i++) {
        for (let j = 1; j < lenWord2; j++) {
            const cost = word1[i - 1] === word2[j - 1] ? 0 : 1;
            matrix[i][j] = Math.min(
                matrix[i - 1][j] + 1,
                matrix[i][j - 1] + 1,
                matrix[i - 1][j - 1] + cost
            );
        }
    }

    return matrix[lenWord1 - 1][lenWord2 - 1];
}

function findClosestMatch(inputWord, wordList) {
    let closestWord = null;
    let minDistance = Infinity;

    for (const word of wordList) {
        const distance = levenshteinDistance(inputWord, word);
        if (distance < minDistance) {
            minDistance = distance;
            closestWord = word;
        }
    }

    return closestWord;
}

// Example usage
const inputWord = "kittn";
const wordList = ["kitten"];

const closestMatch = findClosestMatch(inputWord, wordList);
console.log(`The closest match to '${inputWord}' in the list is '${closestMatch}'`);
